### DISCUSSION TITLE
Potential Wireframes

### DISCUSSION THREAD

**Discussion by christineiym** [gbishop/OS-DPI#15] [Ideas]
As we continue to discuss accessibility and reconfigure the interface, here is a sketch of how an editor interface similar to Microsoft Excel might look:
<img width="820" alt="image" src="https://user-images.githubusercontent.com/69741280/193707243-304f3df7-204c-4bd9-993e-ae4de55e21ee.png">

I am going to consider what goes in the tabs further before taking the wireframe further, so tabs and their contents will definitely change!
I'm also still weighing the pros/cons of having the components as a sidebar versus as a floating window. On a MacBook Pro 14" canvas, there is extra space on the side, but to accommodate multiple screen widths, perhaps a sidebar in the style of the Design/Animations sidebar that pops up in PPT might be better. I wonder what people would prefer?

**Comment by gbishop** [comment]
I'm beginning to think about the low-level machinery we will need to make keyboard access work. 

- I think we need a way to go from DOM node to Component and back (that is, we should be able to jump to the TreeBase component from the generated node in the HTML). 
- We need to keep track of the focus on each tab so when we return to the tab focus is where we left it.
- We need to persist the "open" attribute on any disclosure elements (details/summary) we use so that they remain open through tab changes.

I may try to mock something up that lets me jump to the tab control buttons, choose a tab and restore focus on that tab. Not a menu/toolbar but proof of concept that we have the right low-level functions.

**Comment by christineiym** [comment]
I created a new branch from the [treebase-everywhere branch](https://github.com/gbishop/OS-DPI/tree/treebase-everywhere): [christine-menu-bar](https://github.com/gbishop/OS-DPI/tree/christine-menu-bar) adding tabs to the top toolbar. (It seemed simpler to incorporate it directly into the IDE since most of the functionalities and classes have already been imported). Some notes:

- My focus was on getting tabs into the interface and writing the new [developer's guide](https://github.com/christineiym/OS-DPI/wiki), but I hope to begin exploring the low-level machinery for keyboard access next.
- While the menu toolbar lists sample functionalities, the only buttons that currently work are those leading to guides under Help. 
- There is no search bar, but I plan to add one at the top. I'm thinking if I should make a table mapping functionalities to their onclick actions, keyboard mappings, and subsections/tabs. Perhaps the menu tabs would draw from the table to generate the menu. More importantly, the search bar could draw on the table to display functionalities by fuzzy matching.
- Sorry about the margins; I will need to fix them.
- I copied the existing machinery for TabControl/TabPanel to ensure everything was still connected to TreeBase for logging and accessibility.
- The guides for general use of the interface and for contribution to the source code are separate because I am not a member of UNC-Project-Open-AAC, though I think there are pros and cons to keeping the guides separate. Perhaps if they could be one guide with two sections?
- I moved files with classes extending TabPanel into their own folders (and adjusted imports as necessary): [designer-tabs](https://github.com/gbishop/OS-DPI/tree/christine-menu-bar/components/designer-tabs) and [toolbar-tabs](https://github.com/gbishop/OS-DPI/tree/christine-menu-bar/components/toolbar-tabs). I did this mostly for my benefit in understanding the repository organization, though I am not sure if this is an organizational pattern we would like to continue moving forward.

**Reply by gbishop** [reply]
The code is horribly organized. It wasn't planned, it just happened. I've been hacking on it today trying to eliminate circular imports. Wow. It's a pain.

I think having the developer's guide separate is right.

Maybe instead of a table we could augment the classes with (say) static members that describe what they support? Then (I theorize) the toolbar could query the currently displaying components to learn what it needed to display... I don't know if that would work but I like the idea of localizing the interface to components.

No worries about appearance. 

We can fix your membership but I can't remember how...

I'll take a look at all this tomorrow or Monday, I'm too tired now.

**Reply by christineiym** [reply]
Thank you so much for the feedback! Sounds good. I will reflect more on what you said about the toolbar.
Also, apologies for all the messages over the weekend; that was not intended, I just usually have the most time to complete my work on the weekends. I hope you also get some rest!

**Reply by gbishop** [reply]
Let's talk about this tomorrow when we meet.

**Comment by gbishop** [comment]
Weekend messages are fine. I worked on it off and on today. I work on it
when I feel like working. I switch between programming, thinking about
programming, walking the dog, baking bread and resting.

gb

**Comment by gbishop** [comment]
How about thinking of it as a context menu? Mouse users would right click on a target to choose an action. Keyboard users would have a key sequence to activate it. 
I don't think it changes a lot to think about it as a context menu but it feels a little different to me. 
I guess the key idea here is of "context". You can't just say "Add" you've got to identify the context you're operating in. Right click is a common way to do that.

**Comment by christineiym** [comment]
Right click is an interesting way to think of it! I think it makes sense as long as we have actions in plain sight.

I've begun to implement the interface for the contextual top menus in the [christine-horizontal-menu](https://github.com/gbishop/OS-DPI/tree/christine-horizontal-menu) branch, but it is not complete. It does have the following changes/additions:

- a [toolbarActions class](https://github.com/gbishop/OS-DPI/blob/christine-horizontal-menu/components/toolbar/toolbar-actions.js)
- no more tab panels in the [toolbar](https://github.com/gbishop/OS-DPI/blob/christine-horizontal-menu/components/toolbar/toolbar.js); I'm still working out the TreeBase props to get it to display properly as an unordered list
- minor organizational edits; I did get delayed by import issues (moving the designing view components (i.e., monitor, UI, ...) to a  separate folder disrupted the order in which things were initialized in that TabPanel was not initialized when these components were being initialized) that I made sure to resolve before moving on.

**Reply by christineiym** [reply]
The toolbar also incorporates the menu component, but I was focusing on getting the larger context-specific filtering functionality to work and have not been able to get the dropdown to show up in the proper place yet or to connect the items to callbacks.

**Comment by gbishop** [comment]
In the treebase-everywhere branch I've got something like context menu working in the method getMenuActions. Should it climb the tree listing the actions? For example, when I'm focused on a GridFilter (bad name) the only actions I have are deleting and moving it. But that is a perfectly natural place to add another GridFilter but that is only available on the parent node. If I climb the tree, I could include adding more GridFilters.

Maybe only climb one level? If you go all the way up you're going to have a ton of buttons some of which are point to adding things to far away parents. Stacks get nested for example... If you're in a stack you likely do not want the action from a containing stack. Doesn't seem like you'd even want the actions from your parent in that case.

It's beginning to feel like we need to tweak the included actions on a per component basis.

If a node can't have children then including the parent's actions seem reasonable. When else?

**Comment by christineiym** [comment]
Here are some sketches detailing what I was thinking earlier. A couple notes:

- I apologize in advance for the bad highlighting contrast; the tabs highlighted green say (in order) 'Component Actions', 'Add', 'Component', and '+'.
- The subheading 'Hierarchical Information' refers to the order in which menu actions are categorized.

![20221021_161809814](https://user-images.githubusercontent.com/69741280/197282499-cce64f10-00a1-4694-9819-359ca7791518.jpg)

**Comment by gbishop** [comment]
Excellent. Did you draw these and photographed them?

It would help me understand the implications of the choices if you would
list the keyboard and mouse events necessary to trigger an action.

I'm not convinced that separating parent and child actions is helpful. I
fear that users might be confused by that.

How would these approaches compare to a straightforward menu bar? For
example this one in the w3c accessibility guidelines
<https://www.w3.org/WAI/ARIA/apg/example-index/menubar/menubar-editor.html>.

[image: image.png]

**Reply by christineiym** [reply]
thanks! I drew them and used a scanning app to photograph them.

Apologies for forgetting to list the events! For the sketches, the only events are either (1) a mouse event where the user clicks on the tabs/buttons highlighted green, or (2) a keyboard event where the user presses enter on the appropriate tab after navigating to the tab, taking focus into the sub-menu.

What you say about the parent and child actions makes sense! I suppose that is why I was considering listing them all together without demarcation in the last option.

I think that the last option is most similar to a straightforward menu bar. I do like the simplicity of the menu bar and so am happy to implement it first. I think the only potential drawback is that visual users will not realize that the action lists of the menu are context-dependent until they expand the menu, which I could find personally unexpected, but on second thought could instead be helpful as a form of guidance that takes just a little more effort to discover.

**Comment by gbishop** [comment]
What I meant was the sequence of events to trigger a change. Suppose you're
on the Actions tab and you want to add a condition to the current action.

Consider the menu bar.

WIth the mouse you'd click on the Action to establish the context, then on
the Add menu, then on the Condition entry. Or you might right click on the
action to get the context menu and choose Add condition.

With the keyboard, you'd navigate to the action, Alt (or whatever ) to
focus on the menu bar, A to jump to the Add menu, C to jump to the
Condition entry, Enter to activate, then focus goes to Condition.

 gb

**Reply by christineiym** [reply]
Oh, okay! 

In the 'tab'-menu design, I view the actions displayed as independent of the order in which focus is given (to the tab and to the component). Instead, they update to reflect the last-selected tab / component combination. So either of the following workflows are valid for adding a condition to the current action:

Mouse:

- **Option 1: Component, then tab:** Click on the Action. Assuming you are on the Delete menu, options for deleting the Action will appear. Click on the Add menu, then choose Condition from the list of entries.
- **Option 2: Tab, then component:** Click on the Add menu. Assuming you are not focused on the Action, the list of actions will reflect the last focused on item in the IDE (be it the Actions tab (should Add Action be displayed here?) or a Stack component in the layout tree). Click on the Action, then choose Condition from the updated list of entries. 

Keyboard: 

- **Option 1: Component, then tab:** Tab (or use another key combination - maybe Alt then T?) to the Actions tab and navigate to the Action. Focus on the menu bar with Alt, and type A to jump to the Add menu. Type C to jump to the condition entry, press enter to activate, and focus goes to the Condition.
- **Option 2: Tab, then component:** Same as above, only the the extra step of jumping to the Add menu beforehand, which is unnecessary since the same keyboard events are repeated to jump back to the Add menu to add the Condition after focusing on the appropriate Action. 

(I know you said everything was a state machine, but this especially brings back memories of drawing DFAs, not that we applied them to UI/UX.)
